File: app-service-comments/src/test/java/com/github/arhor/aws/graphql/federation/comments/api/graphql/datafetcher/FederatedEntityFetcherTest.java
package com.github.arhor.aws.graphql.federation.comments.api.graphql.datafetcher;

import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.POST;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.USER;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.Post;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.User;
import com.github.arhor.aws.graphql.federation.comments.service.PostService;
import com.github.arhor.aws.graphql.federation.comments.service.UserService;
import com.netflix.graphql.dgs.DgsQueryExecutor;
import com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration;
import com.netflix.graphql.dgs.autoconfig.DgsExtendedScalarsAutoConfiguration;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@SpringBootTest(
    classes = {
        DgsAutoConfiguration.class,
        DgsExtendedScalarsAutoConfiguration.class,
        FederatedEntityFetcher.class,
    }
)
class FederatedEntityFetcherTest {

    @MockBean
    private UserService userService;

    @MockBean
    private PostService postService;

    @Autowired
    private DgsQueryExecutor dgsQueryExecutor;

    @Test
    void should_create_new_user_representation_for_the_given_id() {
        // Given
        final var userId = 1L;
        final var expectedUser = User.newBuilder().id(userId).build();

        when(userService.findInternalUserRepresentation(any()))
            .thenReturn(expectedUser);

        // When
        var result = dgsQueryExecutor.executeAndExtractJsonPathAsObject(
            """
                query ($representations: [_Any!]!) {
                    _entities(representations: $representations) {
                        ... on User {
                            id
                        }
                    }
                }""".stripIndent(),
            "$.data._entities[0]",
            Map.of("representations", List.of(Map.of("__typename", USER.TYPE_NAME, USER.Id, userId))),
            User.class
        );

        // Then
        assertThat(result)
            .isNotNull()
            .isEqualTo(expectedUser);
    }

    @Test
    void should_create_new_post_representation_for_the_given_id() {
        // Given
        final var postId = 1L;
        final var expectedPost = Post.newBuilder().id(postId).build();

        when(postService.findInternalPostRepresentation(any()))
            .thenReturn(expectedPost);

        // When
        var result = dgsQueryExecutor.executeAndExtractJsonPathAsObject(
            """
                query ($representations: [_Any!]!) {
                    _entities(representations: $representations) {
                        ... on Post {
                            id
                        }
                    }
                }""".stripIndent(),
            "$.data._entities[0]",
            Map.of("representations", List.of(Map.of("__typename", POST.TYPE_NAME, POST.Id, postId))),
            Post.class
        );

        // Then
        assertThat(result)
            .isNotNull()
            .isEqualTo(expectedPost);
    }
}

File: app-service-posts/src/test/kotlin/com/github/arhor/aws/graphql/federation/posts/api/graphql/datafetcher/FederatedEntityFetcherTest.kt
package com.github.arhor.aws.graphql.federation.posts.api.graphql.datafetcher

import com.github.arhor.aws.graphql.federation.posts.generated.graphql.DgsConstants.USER
import com.github.arhor.aws.graphql.federation.posts.generated.graphql.types.User
import com.github.arhor.aws.graphql.federation.posts.service.UserRepresentationService
import com.netflix.graphql.dgs.DgsQueryExecutor
import com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration
import com.netflix.graphql.dgs.autoconfig.DgsExtendedScalarsAutoConfiguration
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.util.UUID

@SpringBootTest(
    classes = [
        DgsAutoConfiguration::class,
        DgsExtendedScalarsAutoConfiguration::class,
        FederatedEntityFetcher::class,
    ]
)
class FederatedEntityFetcherTest {

    @MockkBean
    private lateinit var userRepresentationService: UserRepresentationService

    @Autowired
    private lateinit var dgsQueryExecutor: DgsQueryExecutor

    @Test
    fun `should create new user representation for the given id`() {
        // Given
        val userId = UUID.randomUUID()
        val expectedUser = User(id = userId)

        every { userRepresentationService.findUserRepresentation(any()) } returns expectedUser

        // When
        val result = dgsQueryExecutor.executeAndExtractJsonPathAsObject(
            """
                query (${'$'}representations: [_Any!]!) {
                    _entities(representations: ${'$'}representations) {
                        ... on User {
                            id
                        }
                    }
                }""".trimIndent(),
            "$.data._entities[0]",
            mapOf("representations" to listOf(mapOf("__typename" to USER.TYPE_NAME, USER.Id to userId))),
            User::class.java
        )

        // Then
        verify(exactly = 1) { userRepresentationService.findUserRepresentation(userId) }

        assertThat(result)
            .isNotNull()
            .isEqualTo(expectedUser)
    }
}
